### ISSUE TITLE
Integer index interpreted as float in parallel mode

### ISSUE THREAD

**Issue by polakowo** [numba/numba#7382]
## Reporting a bug

- [ x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

The following snippet

```python
import numpy as np
from numba import njit, prange

@njit
def flex_choose_i_and_col_nb(a, flex_2d):
    i = -1
    col = -1
    if a.ndim == 0:
        i = 0
        col = 0
    elif a.ndim == 1:
        if flex_2d:
            i = 0
            if a.shape[0] == 1:
                col = 0
        else:
            col = 0
            if a.shape[0] == 1:
                i = 0
    else:
        if a.shape[0] == 1:
            i = 0
        if a.shape[1] == 1:
            col = 0
    return i, col


@njit
def flex_select_nb(a, i, col, flex_i, flex_col, flex_2d):
    if flex_i == -1:
        flex_i = i
    if flex_col == -1:
        flex_col = col
    if a.ndim == 0:
        return a.item()
    if a.ndim == 1:
        if flex_2d:
            return a[flex_col]
        return a[flex_i]
    return a[flex_i, flex_col]


@njit
def flex_select_auto_nb(a, i, col, flex_2d):
    flex_i, flex_col = flex_choose_i_and_col_nb(a, flex_2d)
    return flex_select_nb(a, i, col, flex_i, flex_col, flex_2d)

@njit(parallel=True)
def f(shape, a):
    b = np.empty(shape, dtype=np.float_)
    for col in prange(shape[1]):
        for i in range(shape[0]):
            b[i, col] = flex_select_auto_nb(a, i, col, True)
    return b

f((100, 100), np.random.uniform(size=100))
```

fails with 

```plaintext
TypingError: Failed in nopython mode pipeline (step: nopython mode backend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function getitem>) found for signature:
 
 >>> getitem(array(float64, 1d, C), float64)
 
There are 22 candidate implementations:
      - Of which 20 did not match due to:
      Overload of function 'getitem': File: <numerous>: Line N/A.
        With argument(s): '(array(float64, 1d, C), float64)':
       No match.
      - Of which 2 did not match due to:
      Overload in function 'GetItemBuffer.generic': File: numba/core/typing/arraydecl.py: Line 162.
        With argument(s): '(array(float64, 1d, C), float64)':
       Rejected as the implementation raised a specific error:
         TypeError: unsupported array index type float64 in [float64]
  raised from /Users/olegpolakow/miniconda3/lib/python3.7/site-packages/numba/core/typing/arraydecl.py:69

During: typing of intrinsic-call at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (38)

File "../../../../../var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py", line 38:
<source missing, REPL/exec in use?>

During: resolving callee type: type(CPUDispatcher(<function flex_select_nb at 0x7f9ec672ba60>))
During: typing of call at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (46)

During: resolving callee type: type(CPUDispatcher(<function flex_select_nb at 0x7f9ec672ba60>))
During: typing of call at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (46)


File "../../../../../var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py", line 46:
<source missing, REPL/exec in use?>

During: resolving callee type: type(CPUDispatcher(<function flex_select_auto_nb at 0x7f9ec672bae8>))
During: typing of call at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (53)

During: resolving callee type: type(CPUDispatcher(<function flex_select_auto_nb at 0x7f9ec672bae8>))
During: typing of call at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (53)


File "../../../../../var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py", line 53:
<source missing, REPL/exec in use?>

During: lowering "id=14[LoopNest(index_variable = parfor_index.618, range = (0, $26binary_subscr.3, 1))]{34: <ir.Block at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (51)>, 78: <ir.Block at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (53)>, 50: <ir.Block at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (52)>, 52: <ir.Block at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (52)>}Var(parfor_index.618, 2169373933.py:51)" at /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py (51)
```

while it works for `parallel=False`. 

The inspector shows

```plaintext
flex_select_nb (array(float64, 1d, C), int64, int64, int64, int64, Literal[bool](True))
--------------------------------------------------------------------------------
# File: /var/folders/55/jftmp_nx5sg_yyj1xcbrlj1m0000gn/T/ipykernel_66080/2169373933.py
# --- LINE 28 --- 

@njit

# --- LINE 29 --- 

def flex_select_nb(a, i, col, flex_i, flex_col, flex_2d):

    # --- LINE 30 --- 
    # label 0
    #   a = arg(0, name=a)  :: array(float64, 1d, C)
    #   i = arg(1, name=i)  :: int64
    #   col = arg(2, name=col)  :: int64
    #   flex_i = arg(3, name=flex_i)  :: int64
    #   flex_col = arg(4, name=flex_col)  :: int64
    #   flex_2d = arg(5, name=flex_2d)  :: Literal[bool](True)
    #   $const4.1 = const(int, -1)  :: Literal[int](-1)
    #   $6compare_op.2 = flex_i == $const4.1  :: bool
    #   del $const4.1
    #   bool8 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $8pred = call bool8($6compare_op.2, func=bool8, args=(Var($6compare_op.2, 2169373933.py:30),), kws=(), vararg=None)  :: (bool,) -> bool
    #   del bool8
    #   del $6compare_op.2
    #   branch $8pred, 10, 14

    if flex_i == -1:

        # --- LINE 31 --- 
        # label 10
        #   del $8pred
        #   flex_i.1 = i  :: int64
        #   del i

        flex_i = i

    # --- LINE 32 --- 
    #   jump 14
    # label 14
    #   $const16.1 = const(int, -1)  :: Literal[int](-1)
    #   $18compare_op.2 = flex_col == $const16.1  :: bool
    #   del $const16.1
    #   bool20 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $20pred = call bool20($18compare_op.2, func=bool20, args=(Var($18compare_op.2, 2169373933.py:32),), kws=(), vararg=None)  :: (bool,) -> bool
    #   del bool20
    #   del $18compare_op.2
    #   branch $20pred, 22, 26

    if flex_col == -1:

        # --- LINE 33 --- 
        # label 22
        #   del $20pred
        #   flex_col.1 = col  :: int64
        #   del col

        flex_col = col

    # --- LINE 34 --- 
    #   jump 26
    # label 26
    #   $28load_attr.1 = const(int, 1)  :: Literal[int](1)
    #   del $28load_attr.1
    #   $const30.2 = const(int, 0)  :: Literal[int](0)
    #   del $const30.2
    #   $32compare_op.3 = const(int, 0)  :: Literal[int](0)
    #   bool34 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $34pred = call bool34($32compare_op.3, func=bool34, args=(Var($32compare_op.3, 2169373933.py:34),), kws=(), vararg=None)  :: (int64,) -> bool
    #   del bool34
    #   del $34pred
    #   del $32compare_op.3
    #   jump 44

    if a.ndim == 0:

        # --- LINE 35 --- 

        return a.item()

    # --- LINE 36 --- 
    # label 44
    #   $46load_attr.1 = const(int, 1)  :: Literal[int](1)
    #   del $46load_attr.1
    #   $const48.2 = const(int, 1)  :: Literal[int](1)
    #   del $const48.2
    #   $50compare_op.3 = const(int, 1)  :: Literal[int](1)
    #   bool52 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $52pred = call bool52($50compare_op.3, func=bool52, args=(Var($50compare_op.3, 2169373933.py:36),), kws=(), vararg=None)  :: (int64,) -> bool
    #   del bool52
    #   del $52pred
    #   del $50compare_op.3
    #   jump 54

    if a.ndim == 1:

        # --- LINE 37 --- 
        # label 54
        #   bool56 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
        #   $56pred = call bool56(flex_2d, func=bool56, args=(Var(flex_2d, 2169373933.py:30),), kws=(), vararg=None)  :: (bool,) -> bool
        #   del flex_2d
        #   del bool56
        #   branch $56pred, 58, 66

        if flex_2d:

            # --- LINE 38 --- 
            #   del col
            #   del $20pred
            #   flex_col.2 = phi(incoming_values=[Var(flex_col, 2169373933.py:30), Var(flex_col.1, 2169373933.py:33)], incoming_blocks=[14, 22])  :: int64
            #   del flex_col.1
            #   del flex_col
            # label 58
            #   del flex_i.2
            #   del $56pred
            #   $62binary_subscr.2 = getitem(value=a, index=flex_col.2, fn=<built-in function getitem>)  :: float64
            #   del flex_col.2
            #   del a
            #   $64return_value.3 = cast(value=$62binary_subscr.2)  :: float64
            #   del $62binary_subscr.2
            #   return $64return_value.3

            return a[flex_col]

        # --- LINE 39 --- 
        #   del i
        #   del $8pred
        #   flex_i.2 = phi(incoming_values=[Var(flex_i, 2169373933.py:30), Var(flex_i.1, 2169373933.py:31)], incoming_blocks=[0, 10])  :: int64
        #   del flex_i.1
        #   del flex_i
        # label 66
        #   del flex_col.2
        #   del $56pred
        #   $70binary_subscr.2 = getitem(value=a, index=flex_i.2, fn=<built-in function getitem>)  :: float64
        #   del flex_i.2
        #   del a
        #   $72return_value.3 = cast(value=$70binary_subscr.2)  :: float64
        #   del $70binary_subscr.2
        #   return $72return_value.3

        return a[flex_i]

    # --- LINE 40 --- 

    return a[flex_i, flex_col]
```

So nothing unusual.

The column index `col` is always an integer, but in the parallel mode, it suddenly becomes a float. What's interesting is that casting to an integer like this `return a[int(flex_col)]` in `flex_select_nb` solves the issue, while casting it in any other line ignores the casting and throws the same error. Am I missing something about parallelization with Numba? Is there something special about the loop variable in `prange`?

Context: I'm trying to create a function that selects an element out of `a` as if it was broadcast, so I can avoid wasting memory on broadcasting. 

Edit: tested both on 0.53.1 and 0.54.0

**Comment by sklam** [comment]
This is a known problem with the casting rule borrowed from `numpy` for `int64` and `uint64`. 

```python
>>> import numpy as np
>>> (np.int64(1) + np.uint64(0)).dtype
dtype('float64')
```

`prange` may yield `uint64` due to optimization reasons. 

```python
@njit
def flex_select_nb(a, i, col, flex_i, flex_col, flex_2d):
    if flex_i == -1:
        flex_i = i
    if flex_col == -1:
        flex_col = col    # if col is uint64 and flex_col is int64, flex_col is forced to float64.
    if a.ndim == 0:
        return a.item()
    if a.ndim == 1:
        if flex_2d:
            return a[flex_col]
        return a[flex_i]
    return a[flex_i, flex_col]
```

A workaround for this problem is explicitly casting `col` for `int64`; for example:

```python
flex_col = np.int64(col)
```

**Comment by sklam** [comment]
Duplicate of https://github.com/numba/numba/issues/5801

**Comment by polakowo** [comment]
Got it, thanks @sklam
